File: client/src/types/shipment-event.ts
import { z } from "zod";
import { userSchema } from "./user";

export const shipmentEventTypeSchema = z.enum([
  "ShipmentCreated",
  "ShipmentUpdated",
  "StatusChanged",
  "ShipmentCanceled",
  "ShipmentUncanceled",
  "OwnershipTransferred",
  "MoveStatusChanged",
  "MoveDeparted",
  "MoveArrived",
  "StopCompleted",
  "DriverAssigned",
  "DriverReassigned",
  "DriverUnassigned",
  "HoldPlaced",
  "HoldUpdated",
  "HoldReleased",
  "CommentPosted",
]);
export type ShipmentEventType = z.infer<typeof shipmentEventTypeSchema>;

export const shipmentEventSeveritySchema = z.enum([
  "danger",
  "success",
  "brand",
  "info",
  "muted",
]);
export type ShipmentEventSeverity = z.infer<typeof shipmentEventSeveritySchema>;

export const shipmentEventActorTypeSchema = z.enum(["user", "apikey", "system", "edi"]);
export type ShipmentEventActorType = z.infer<typeof shipmentEventActorTypeSchema>;

export const shipmentEventSchema = z.object({
  id: z.string(),
  organizationId: z.string(),
  businessUnitId: z.string(),
  shipmentId: z.string(),
  moveId: z.string().optional(),
  stopId: z.string().optional(),
  assignmentId: z.string().optional(),
  commentId: z.string().optional(),
  holdId: z.string().optional(),
  type: shipmentEventTypeSchema,
  severity: shipmentEventSeveritySchema,
  actorType: shipmentEventActorTypeSchema,
  actorId: z.string().optional(),
  actorLabel: z.string().default(""),
  summary: z.string(),
  metadata: z.record(z.string(), z.unknown()).default({}),
  occurredAt: z.number(),
  correlationId: z.string().optional(),
  actor: userSchema.partial().optional(),
  shipment: z
    .object({
      id: z.string(),
      proNumber: z.string(),
    })
    .partial()
    .optional(),
});

export type ShipmentEvent = z.infer<typeof shipmentEventSchema>;

export const shipmentEventListSchema = z.array(shipmentEventSchema);
export type ShipmentEventList = z.infer<typeof shipmentEventListSchema>;

File: services/tms/internal/core/ports/services/shipmentevent.go
package services

import (
	"context"

	"github.com/emoss08/trenova/internal/core/domain/shipmentevent"
	"github.com/emoss08/trenova/internal/core/ports/repositories"
	"github.com/emoss08/trenova/shared/pulid"
)

type RecordShipmentEventParams struct {
	OrganizationID pulid.ID
	BusinessUnitID pulid.ID
	ShipmentID     pulid.ID
	MoveID         pulid.ID
	StopID         pulid.ID
	AssignmentID   pulid.ID
	CommentID      pulid.ID
	HoldID         pulid.ID

	Type     shipmentevent.Type
	Severity shipmentevent.Severity
	Summary  string
	Metadata map[string]any

	Actor         AuditActor
	ActorLabel    string
	CorrelationID string
	OccurredAt    int64
}

type ShipmentEventService interface {
	Record(ctx context.Context, params *RecordShipmentEventParams) error
	List(
		ctx context.Context,
		req *repositories.ListShipmentEventsRequest,
	) ([]*shipmentevent.Event, error)
}

type ShipmentEventObserver interface {
	OnShipmentEvent(ctx context.Context, event *shipmentevent.Event) error
}
